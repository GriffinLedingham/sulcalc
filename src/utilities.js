export const Gens = {
  RBY: 1,
  GSC: 2,
  ADV: 3,
  HGSS: 4,
  B2W2: 5,
  ORAS: 6,
  SM: 7
};

export const maxGen = Math.max(...Object.values(Gens));

export const Stats = {
  HP: 0,
  ATK: 1,
  DEF: 2,
  SATK: 3,
  SDEF: 4,
  SPD: 5,
  ACC: 6,
  EVA: 7,
  SPC: 3
};

export const Genders = {
  NO_GENDER: 0,
  MALE: 1,
  FEMALE: 2
};

export const DamageClasses = {
  OTHER: 0,
  PHYSICAL: 1,
  SPECIAL: 2
};

export const Statuses = {
  NO_STATUS: 0,
  POISONED: 1,
  BADLY_POISONED: 2,
  BURNED: 3,
  PARALYZED: 4,
  ASLEEP: 5,
  FROZEN: 6
};

export const Types = {
  NORMAL: 0,
  FIGHTING: 1,
  FLYING: 2,
  POISON: 3,
  GROUND: 4,
  ROCK: 5,
  BUG: 6,
  GHOST: 7,
  STEEL: 8,
  FIRE: 9,
  WATER: 10,
  GRASS: 11,
  ELECTRIC: 12,
  PSYCHIC: 13,
  ICE: 14,
  DRAGON: 15,
  DARK: 16,
  FAIRY: 17,
  CURSE: 18
};

export const Weathers = {
  CLEAR: 0,
  SUN: 4,
  RAIN: 2,
  SAND: 3,
  HAIL: 1,
  HARSH_SUN: 6,
  HEAVY_RAIN: 5,
  STRONG_WINDS: 7
};

export const Natures = {
  HARDY: 0,
  LONELY: 1,
  BRAVE: 2,
  ADAMANT: 3,
  NAUGHTY: 4,
  BOLD: 5,
  DOCILE: 6,
  RELAXED: 7,
  IMPISH: 8,
  LAX: 9,
  TIMID: 10,
  HASTY: 11,
  SERIOUS: 12,
  JOLLY: 13,
  NAIVE: 14,
  MODEST: 15,
  MILD: 16,
  QUIET: 17,
  BASHFUL: 18,
  RASH: 19,
  CALM: 20,
  GENTLE: 21,
  SASSY: 22,
  CAREFUL: 23,
  QUIRKY: 24
};

export const Terrains = {
  NO_TERRAIN: Symbol("sulcalc.Terrains.NO_TERRAIN"),
  GRASSY_TERRAIN: Symbol("sulcalc.Terrains.GRASSY_TERRAIN"),
  MISTY_TERRAIN: Symbol("sulcalc.Terrains.MISTY_TERRAIN"),
  ELECTRIC_TERRAIN: Symbol("sulcalc.Terrains.ELECTRIC_TERRAIN"),
  PSYCHIC_TERRAIN: Symbol("sulcalc.Terrains.PSYCHIC_TERRAIN")
};

export const roundHalfToZero = n =>
  Math.trunc(n) + (n % 1 > 0.5) - (n % 1 < -0.5);

export const chainMod = (modifier1, modifier2) =>
  (modifier1 * modifier2 + 0x800) >> 12;

export const applyMod = (modifier, value) =>
  roundHalfToZero((value * modifier) / 0x1000);

export const applyModAll = (modifier, values) =>
  values.map(v => applyMod(modifier, v));

export function damageVariation(baseDamage, min, max) {
  const damages = [];
  for (let i = min; i <= max; i++) {
    damages.push(Math.trunc((baseDamage * i) / max));
  }
  return damages;
}

export const needsScaling = (...stats) => stats.some(stat => stat > 255);

export const scaleStat = (stat, bits = 2) => (stat >> bits) & 0xff;
