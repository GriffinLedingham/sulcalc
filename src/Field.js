import { Terrains, Weathers, maxGen } from "./utilities";

export default class Field {
  constructor(field = {}) {
    this.gen = Number(field.gen ?? maxGen);
    this.multiBattle = Boolean(field.multiBattle);
    this.invertedBattle = Boolean(field.invertedBattle);

    this.weather = Number(field.weather ?? Weathers.CLEAR);
    this.airLock = Boolean(field.airLock);

    this.waterSport = Boolean(field.waterSport);
    this.mudSport = Boolean(field.mudSport);

    this.magicRoom = Boolean(field.magicRoom);
    this.wonderRoom = Boolean(field.wonderRoom);
    this.gravity = Boolean(field.gravity);

    this.terrain = field.terrain ?? Terrains.NO_TERRAIN;

    this.fairyAura = Boolean(field.fairyAura);
    this.darkAura = Boolean(field.darkAura);
    this.auraBreak = Boolean(field.auraBreak);

    this.ionDeluge = Boolean(field.ionDeluge);
  }

  effectiveWeather() {
    return this.airLock ? Weathers.CLEAR : this.weather;
  }

  isClearWeather() {
    return this.effectiveWeather() === Weathers.CLEAR;
  }

  rain() {
    return [Weathers.RAIN, Weathers.HEAVY_RAIN].includes(
      this.effectiveWeather()
    );
  }

  sun() {
    return [Weathers.SUN, Weathers.HARSH_SUN].includes(this.effectiveWeather());
  }

  sand() {
    return this.effectiveWeather() === Weathers.SAND;
  }

  hail() {
    return this.effectiveWeather() === Weathers.HAIL;
  }

  heavyRain() {
    return this.effectiveWeather() === Weathers.HEAVY_RAIN;
  }

  harshSun() {
    return this.effectiveWeather() === Weathers.HARSH_SUN;
  }

  strongWinds() {
    return this.effectiveWeather() === Weathers.STRONG_WINDS;
  }

  grassyTerrain() {
    return this.terrain === Terrains.GRASSY_TERRAIN;
  }

  mistyTerrain() {
    return this.terrain === Terrains.MISTY_TERRAIN;
  }

  electricTerrain() {
    return this.terrain === Terrains.ELECTRIC_TERRAIN;
  }

  psychicTerrain() {
    return this.terrain === Terrains.PSYCHIC_TERRAIN;
  }
}
