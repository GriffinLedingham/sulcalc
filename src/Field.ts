import { Generation, Terrain, Weather, maxGen } from "./utilities";

export type FieldOptions = Partial<Field>;

export default class Field {
  public gen: Generation = maxGen;
  public weather: Weather = Weather.CLEAR;
  public terrain: Terrain = Terrain.NO_TERRAIN;
  public multiBattle: boolean = false;
  public invertedBattle: boolean = false;
  public airLock: boolean = false;
  public waterSport: boolean = false;
  public mudSport: boolean = false;
  public magicRoom: boolean = false;
  public wonderRoom: boolean = false;
  public gravity: boolean = false;
  public fairyAura: boolean = false;
  public darkAura: boolean = false;
  public auraBreak: boolean = false;
  public ionDeluge: boolean = false;

  public constructor(field: Field | FieldOptions = {}) {
    Object.assign(this, field);
  }

  public effectiveWeather(): Weather {
    return this.airLock ? Weather.CLEAR : this.weather;
  }

  public isClearWeather(): boolean {
    return this.effectiveWeather() === Weather.CLEAR;
  }

  public rain(): boolean {
    const weather = this.effectiveWeather();
    return weather === Weather.RAIN || weather === Weather.HEAVY_RAIN;
  }

  public sun(): boolean {
    const weather = this.effectiveWeather();
    return weather === Weather.SUN || weather === Weather.HARSH_SUN;
  }

  public sand(): boolean {
    return this.effectiveWeather() === Weather.SAND;
  }

  public hail(): boolean {
    return this.effectiveWeather() === Weather.HAIL;
  }

  public heavyRain(): boolean {
    return this.effectiveWeather() === Weather.HEAVY_RAIN;
  }

  public harshSun(): boolean {
    return this.effectiveWeather() === Weather.HARSH_SUN;
  }

  public strongWinds(): boolean {
    return this.effectiveWeather() === Weather.STRONG_WINDS;
  }

  public grassyTerrain(): boolean {
    return this.terrain === Terrain.GRASSY_TERRAIN;
  }

  public mistyTerrain(): boolean {
    return this.terrain === Terrain.MISTY_TERRAIN;
  }

  public electricTerrain(): boolean {
    return this.terrain === Terrain.ELECTRIC_TERRAIN;
  }

  public psychicTerrain(): boolean {
    return this.terrain === Terrain.PSYCHIC_TERRAIN;
  }
}
