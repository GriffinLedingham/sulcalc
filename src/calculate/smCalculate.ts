import Field from "../Field";
import Move from "../Move";
import Pokemon from "../Pokemon";
import {
  applyMod,
  applyModAll,
  chainMod,
  damageVariation,
  Stat,
  Type
} from "../utilities";
import moveInfo from "./moveInfo";

/*
 * TODO
 * Moves:
 * Shell Trap
 * Guardian of Alola 75% HP
 * Revelation Dance
 * Sparkling Aria heals burns
 * Baneful Bunker poisons
 * Burn Up
 *
 * Abilities:
 * Stakeout
 * Soul-Heart
 * Shields Down
 * Schooling
 */
export default (
  attacker: Pokemon,
  defender: Pokemon,
  move: Move,
  field: Field
): number[] => {
  const info = moveInfo(attacker, defender, move, field);
  if (info.fail || info.effectiveness[0] === 0) return [0];
  const {
    moveType,
    movePower,
    effectiveness,
    superEffective,
    notVeryEffective
  } = info;

  const defStat = field.wonderRoom ? Stat.SDEF : Stat.DEF;
  const sdefStat = field.wonderRoom ? Stat.DEF : Stat.SDEF;
  const unawareA = attacker.ability.name === "Unaware";
  const unawareD = defender.ability.name === "Unaware";
  let atk, def, satk, sdef;
  if (move.name === "Foul Play") {
    if (unawareA) {
      def = defender.stat(defStat);
      sdef = defender.stat(sdefStat);
      atk = defender.stat(Stat.ATK);
    } else if (move.critical) {
      def = Math.min(defender.stat(defStat), defender.boostedStat(defStat));
      sdef = Math.min(defender.stat(sdefStat), defender.boostedStat(sdefStat));
      atk = Math.max(defender.stat(Stat.ATK), defender.boostedStat(Stat.ATK));
    } else {
      def = defender.boostedStat(defStat);
      sdef = defender.boostedStat(sdefStat);
      atk = defender.boostedStat(Stat.ATK);
    }

    if (unawareD) {
      satk = attacker.stat(Stat.SATK);
    } else if (move.critical) {
      satk = Math.max(
        attacker.stat(Stat.SATK),
        attacker.boostedStat(Stat.SATK)
      );
    } else {
      satk = attacker.boostedStat(Stat.SATK);
    }
  } else if (move.ignoresDefenseBoosts()) {
    def = defender.stat(defStat);
    sdef = defender.stat(sdefStat);

    if (unawareD) {
      atk = attacker.stat(Stat.ATK);
      satk = attacker.stat(Stat.SATK);
    } else if (move.critical) {
      atk = Math.max(attacker.stat(Stat.ATK), attacker.boostedStat(Stat.ATK));
      satk = Math.max(
        attacker.stat(Stat.SATK),
        attacker.boostedStat(Stat.SATK)
      );
    } else {
      atk = attacker.boostedStat(Stat.ATK);
      satk = attacker.boostedStat(Stat.SATK);
    }
  } else {
    if (unawareA) {
      def = defender.stat(defStat);
      sdef = defender.stat(sdefStat);
    } else if (move.critical) {
      def = Math.min(defender.stat(defStat), defender.boostedStat(defStat));
      sdef = Math.min(defender.stat(sdefStat), defender.boostedStat(sdefStat));
    } else {
      def = defender.boostedStat(defStat);
      sdef = defender.boostedStat(sdefStat);
    }

    if (unawareD) {
      atk = attacker.stat(Stat.ATK);
      satk = attacker.stat(Stat.SATK);
    } else if (move.critical) {
      atk = Math.max(attacker.stat(Stat.ATK), attacker.boostedStat(Stat.ATK));
      satk = Math.max(
        attacker.stat(Stat.SATK),
        attacker.boostedStat(Stat.SATK)
      );
    } else {
      atk = attacker.boostedStat(Stat.ATK);
      satk = attacker.boostedStat(Stat.SATK);
    }
  }

  let atkMod = 0x1000;
  let satkMod = 0x1000;

  if (
    defender.ability.name === "Thick Fat" &&
    (moveType === Type.FIRE || moveType === Type.ICE)
  ) {
    atkMod = chainMod(0x800, atkMod);
    satkMod = chainMod(0x800, satkMod);
  }

  switch (attacker.ability.name) {
    case "Guts":
      if (defender.status) atkMod = chainMod(0x1800, atkMod);
      break;
    case "Plus":
      if (attacker.minus) satkMod = chainMod(0x1800, satkMod);
      break;
    case "Minus":
      if (attacker.plus) satkMod = chainMod(0x1800, satkMod);
      break;
    case "Defeatist":
      if (attacker.currentHp * 2 <= attacker.stat(Stat.HP)) {
        atkMod = chainMod(0x800, atkMod);
        satkMod = chainMod(0x800, satkMod);
      }
      break;
    case "Huge Power":
    case "Pure Power":
      atkMod = chainMod(0x2000, atkMod);
      break;
    case "Solar Power":
      if (field.sun()) satkMod = chainMod(0x1800, satkMod);
      break;
    case "Hustle":
      atk = applyMod(0x1800, atk);
      break;
    case "Flash Fire":
      if (attacker.flashFire && moveType === Type.FIRE) {
        atkMod = chainMod(0x1800, atkMod);
        satkMod = chainMod(0x1800, satkMod);
      }
      break;
    case "Slow Start":
      if (attacker.slowStart) atkMod = chainMod(0x800, atkMod);
      break;
    default:
      if (attacker.pinchAbilityActivated(moveType)) {
        // blaze, torrent, overgrow, ...
        atkMod = chainMod(0x1800, atkMod);
        satkMod = chainMod(0x1800, satkMod);
      }
  }

  if (attacker.flowerGift && field.sun()) {
    atkMod = chainMod(0x1800, atkMod);
  }

  if (attacker.battery) {
    satkMod = chainMod(0x14cd, satkMod);
  }

  switch (attacker.item.name) {
    case "Deep Sea Tooth":
      if (attacker.name === "Clamperl") {
        satkMod = chainMod(0x2000, satkMod);
      }
      break;
    case "Choice Band":
      atkMod = chainMod(0x1800, atkMod);
      break;
    case "Choice Specs":
      satkMod = chainMod(0x1800, satkMod);
      break;
    default:
      if (attacker.thickClubBoosted()) {
        atkMod = chainMod(0x2000, atkMod);
      } else if (attacker.lightBallBoosted()) {
        atkMod = chainMod(0x2000, atkMod);
        satkMod = chainMod(0x2000, satkMod);
      }
  }

  atk = applyMod(atkMod, atk);
  satk = applyMod(satkMod, satk);

  if (field.sand() && defender.stab(Type.ROCK)) {
    sdef = applyMod(0x1800, sdef);
  }

  let defMod = 0x1000;
  let sdefMod = 0x1000;

  if (defender.ability.name === "Marvel Scale" && defender.status) {
    defMod = chainMod(0x1800, defMod);
  } else if (defender.ability.name === "Grass Pelt" && field.grassyTerrain()) {
    defMod = chainMod(0x1800, defMod);
  }

  if (defender.flowerGift && field.sun()) {
    sdefMod = chainMod(0x1800, sdefMod);
  }

  switch (defender.item.name) {
    case "Deep Sea Scale":
      if (defender.name === "Clamperl") {
        sdefMod = chainMod(0x1800, sdefMod);
      }
      break;
    case "Metal Powder":
      if (defender.name === "Ditto") {
        defMod = chainMod(0x2000, defMod);
      }
      break;
    case "Eviolite":
      if (defender.hasEvolution()) {
        defMod = chainMod(0x1800, defMod);
        sdefMod = chainMod(0x1800, sdefMod);
      }
      break;
    case "Assault Vest":
      sdefMod = chainMod(0x1800, sdefMod);
      break;
    /* no default */
  }

  def = applyMod(defMod, def);
  sdef = applyMod(sdefMod, sdef);

  let a, d;
  if (move.usesMaxAttackingStat()) {
    a = Math.max(atk, satk);
    d = sdef;
  } else if (move.isPsyshockLike()) {
    a = satk;
    d = def;
  } else if (move.isPhysical()) {
    a = atk;
    d = def;
  } else if (move.isSpecial()) {
    a = satk;
    d = sdef;
  } else {
    return [0];
  }

  let baseDamage = Math.trunc(
    Math.trunc((Math.trunc((2 * attacker.level) / 5 + 2) * movePower * a) / d) /
      50
  );
  baseDamage += 2;

  if (field.multiBattle && move.hasMultipleTargets()) {
    baseDamage = applyMod(0xc00, baseDamage);
  }

  if (move.name !== "Weather Ball") {
    if (
      (field.harshSun() && moveType === Type.WATER) ||
      (field.heavyRain() && moveType === Type.FIRE)
    ) {
      return [0];
    }
    if (field.sun()) {
      if (moveType === Type.FIRE) {
        baseDamage = applyMod(0x1800, baseDamage);
      } else if (moveType === Type.WATER) {
        baseDamage = applyMod(0x800, baseDamage);
      }
    } else if (field.rain()) {
      if (moveType === Type.WATER) {
        baseDamage = applyMod(0x1800, baseDamage);
      } else if (moveType === Type.FIRE) {
        baseDamage = applyMod(0x800, baseDamage);
      }
    }
  }

  if (
    ((field.grassyTerrain() && moveType === Type.GRASS) ||
      (field.electricTerrain() && moveType === Type.ELECTRIC) ||
      (field.psychicTerrain() && moveType === Type.PSYCHIC)) &&
    attacker.isGrounded(field)
  ) {
    baseDamage = applyMod(0x1800, baseDamage);
  }

  if (
    field.mistyTerrain() &&
    defender.isGrounded(field) &&
    moveType === Type.DRAGON
  ) {
    baseDamage = applyMod(0x800, baseDamage);
  }

  if (move.critical) {
    baseDamage = applyMod(0x1800, baseDamage);
  }

  let damages = damageVariation(baseDamage, 85, 100);

  if (attacker.stab(moveType) || attacker.ability.name === "Protean") {
    damages =
      attacker.ability.name === "Adaptability"
        ? applyModAll(0x2000, damages)
        : applyModAll(0x1800, damages);
  }

  damages = damages.map(d =>
    Math.trunc((d * effectiveness[0]) / effectiveness[1])
  );

  if (
    attacker.isBurned() &&
    move.isPhysical() &&
    attacker.ability.name !== "Guts" &&
    move.name !== "Facade"
  ) {
    damages = damages.map(d => Math.trunc(d / 2));
  }

  damages = damages.map(d => Math.max(1, d));

  let finalMod = 0x1000;

  if (!move.critical && attacker.ability.name !== "Infiltrator") {
    if (defender.reflect && (move.isPhysical() || move.isPsyshockLike())) {
      finalMod = chainMod(field.multiBattle ? 0xa8f : 0x800, finalMod);
    }
    if (defender.lightScreen && move.isSpecial() && !move.isPsyshockLike()) {
      finalMod = chainMod(field.multiBattle ? 0xa8f : 0x800, finalMod);
    }
  }

  if (defender.multiscaleIsActive()) {
    finalMod = chainMod(0x800, finalMod);
  }

  if (
    defender.ability.name === "Fur Coat" &&
    (move.isPhysical() || move.isPsyshockLike())
  ) {
    finalMod = chainMod(0x800, finalMod);
  }

  if (notVeryEffective && attacker.ability.name === "Tinted Lens") {
    finalMod = chainMod(0x2000, finalMod);
  }

  if (defender.friendGuard) {
    finalMod = chainMod(0xc00, finalMod);
  }

  if (defender.auroraVeil) {
    finalMod = chainMod(field.multiBattle ? 0xaaa : 0x800, finalMod);
  }

  if (attacker.ability.name === "Sniper" && move.critical) {
    finalMod = chainMod(0x1800, finalMod);
  }

  if (superEffective && defender.ability.reducesSuperEffective()) {
    finalMod = chainMod(0xc00, finalMod);
  }

  if (field.grassyTerrain() && move.weakenedByGrassyTerrain()) {
    finalMod = chainMod(0x800, finalMod);
  }

  switch (attacker.item.name) {
    case "Metronome":
      if (attacker.metronome <= 4) {
        const mod = 0x1000 + attacker.metronome * 0x333;
        finalMod = chainMod(mod, finalMod);
      } else {
        finalMod = chainMod(0x2000, finalMod);
      }
      break;
    case "Expert Belt":
      if (superEffective) {
        finalMod = chainMod(0x1333, finalMod);
      }
      break;
    case "Life Orb":
      finalMod = chainMod(0x14cc, finalMod);
      break;
    /* no default */
  }

  if (
    defender.item.berryTypeResist() === moveType &&
    (superEffective || moveType === Type.NORMAL)
  ) {
    finalMod = chainMod(0x800, finalMod);
    defender.item.used = true;
  }

  if (
    (move.dig && move.boostedByDig()) ||
    (move.dive && move.boostedByDive()) ||
    (move.minimize && move.boostedByMinimize())
  ) {
    finalMod = chainMod(0x2000, finalMod);
  }

  damages = applyModAll(finalMod, damages);

  return damages;
};
