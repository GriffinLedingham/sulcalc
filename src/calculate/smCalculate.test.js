import smCalculate from "sulcalc/calculate/smCalculate";
import Pokemon from "sulcalc/Pokemon";
import Move from "sulcalc/Move";
import Field from "sulcalc/Field";
import { Gens, Natures, Terrains } from "sulcalc/utilities";

const gen = Gens.SM;

let field, landorusT, heatran, lavaPlume, earthquake;
beforeEach(() => {
  field = new Field({ gen });
  heatran = new Pokemon({
    name: "Heatran",
    evs: [252, 0, 0, 0, 4, 252],
    nature: Natures.TIMID,
    gen
  });
  landorusT = new Pokemon({
    name: "Landorus-Therian",
    evs: [252, 0, 216, 0, 24, 16],
    nature: Natures.IMPISH,
    gen
  });
  lavaPlume = new Move({ name: "Lava Plume", gen });
  earthquake = new Move({ name: "Earthquake", gen });
});

test("sanity check", () => {
  const damage = smCalculate(heatran, landorusT, lavaPlume, field);
  expect(Math.max(...damage)).toEqual(150);
});

test("Aurora Veil", () => {
  heatran.auroraVeil = true;
  const damageSingle = smCalculate(landorusT, heatran, earthquake, field);
  expect(damageSingle).toMatchSnapshot();

  field.multiBattle = true;
  const damageMulti = smCalculate(landorusT, heatran, earthquake, field);
  expect(damageMulti).toMatchSnapshot();

  expect(damageSingle).not.toEqual(damageMulti);
});

test("Fairy Aura is a move power mod", () => {
  const xerneas = new Pokemon({
    name: "Xerneas",
    item: "Life Orb",
    nature: Natures.MODEST,
    evs: [0, 0, 0, 252, 0, 0],
    gen
  });
  const genesect = new Pokemon({
    name: "Genesect",
    nature: Natures.NAIVE,
    gen
  });
  const moonblast = new Move({ name: "Moonblast", gen });
  const auraField = new Field({ fairyAura: true, gen });
  expect(
    smCalculate(xerneas, genesect, moonblast, auraField)
  ).toMatchSnapshot();
});

test("Neuroforce boosts super effective attacks", () => {
  const necrozma = new Pokemon({ name: "Necrozma-Dawn-Wings", gen });
  const surf = new Move({ name: "Surf", gen });
  const damage = smCalculate(necrozma, heatran, surf, field);
  necrozma.ability.name = "Neuroforce";
  const boostedDamage = smCalculate(necrozma, heatran, surf, field);
  expect(Math.max(...boostedDamage)).toBeGreaterThan(Math.max(...damage));
});

test("Psychic Terrain blocks priority", () => {
  const quickAttack = new Move({ name: "Quick Attack", gen });
  const psychicTerrain = new Field({
    terrain: Terrains.PSYCHIC_TERRAIN,
    gen
  });
  const damage = smCalculate(heatran, landorusT, quickAttack, psychicTerrain);
  expect(damage).not.toEqual([0]);

  landorusT.item.name = "Iron Ball";
  const notDamage = smCalculate(
    heatran,
    landorusT,
    quickAttack,
    psychicTerrain
  );
  expect(notDamage).toEqual([0]);
});

test("Knock Off should multiply base power by 1.5x if it gets rid of an item", () => {
  const knockOff = new Move({ name: "Knock Off", gen });
  const bisharp = new Pokemon({ name: "Bisharp", gen });
  const rotomW = new Pokemon({ name: "Rotom-Wash", gen });
  const damage = smCalculate(bisharp, rotomW, knockOff, field);
  expect(Math.max(...damage)).toEqual(96);
  rotomW.item.name = "Rocky Helmet";
  const boostedDamage = smCalculate(bisharp, rotomW, knockOff, field);
  expect(Math.max(...boostedDamage)).toEqual(142);
});
