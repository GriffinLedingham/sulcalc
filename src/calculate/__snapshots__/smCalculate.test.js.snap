// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smCalculate() Aurora Veil 1`] = `
Array [
  284,
  288,
  290,
  294,
  296,
  300,
  302,
  308,
  312,
  314,
  318,
  320,
  324,
  326,
  330,
  336,
]
`;

exports[`smCalculate() Aurora Veil 2`] = `
Array [
  283,
  288,
  291,
  291,
  296,
  299,
  304,
  307,
  312,
  312,
  315,
  320,
  323,
  328,
  331,
  336,
]
`;

exports[`smCalculate() Fairy Aura is a move power mod 1`] = `
Array [
  172,
  173,
  175,
  178,
  179,
  182,
  183,
  186,
  187,
  190,
  191,
  194,
  195,
  198,
  199,
  203,
]
`;
